Order book

EvgeniiaVak opened the pull request:
closes #32 

- consolidates Orderbook-related elements in single .tsx component
- adds spread
- adds total column

fliebenberg commented:
This is looking great @EvgeniiaVak . Can I make one suggestion: Can you make the Price column display the numbers with a fixed number of decimals (the max digits for token2 from the pairInfo). That way the prices display nicely under each other and your eye does not have to make the adjustment for longer and shorter prices.
We will eventually also want to display the other numbers with commas/spaces between thousands etc, so not sure if you want to work on number formatting functionality in general while you are at it.

fliebenberg commented:
@EvgeniiaVak I have added a PR where I added a utility file to the code with some functions for displaying formatted numbers. You can use that to easily format the price field as suggested.

SmashingBumpkin reviewed (approved):
Approved on the assumption it's easiest to merge this, then merge the utils in to apply rounding afterwards.

EvgeniiaVak commented:
Please don't merge (I cannot make it draft for some reason), I'm already fixing the conflicts by importing the utils to branch to this one. Also will remove the TODO link.

EvgeniiaVak commented on src/app/OrderBook.tsx:
```diff
@@ -2,29 +2,130 @@ import { useContext } from "react";
+        </tr>
+        <tr className="border-none">
+          <td className="text-sm m-0 p-0" colSpan={4}>
+            TODO: should we use USD value?
```
yeah, there is a discussion about that in figma: https://www.figma.com/file/P7pfzKwJ4G6ClapXfl61D3?node-id=18:3324&mode=design#505524313

EvgeniiaVak commented on src/app/OrderBook.tsx:
```diff
@@ -2,29 +2,130 @@ import { useContext } from "react";
+
+  if (bestBuy && bestSell && bestBuy !== 0) {
+    const spread = bestSell - bestBuy;
+    const spreadPercent = (spread / bestBuy) * 100;
```
I'll include using Fred's utils functions for rounding.

"Do they update automatically when orders get bought?" - they all depend on the adex state, so I guess yes. I'll try to test this.

About the `spreadPercent` I don't really know... maybe 🤔 I just took Fred's word for it (cc @fliebenberg ), which was:
> It is indeed the difference between the best buy and sell prices. You already have this info in the orderbook, so the frontend just needs to calculate the difference between the two values. It can then also calculate the % spread, which is just the spread / best buy price.

from https://www.figma.com/file/P7pfzKwJ4G6ClapXfl61D3?node-id=18:3324&mode=design#505524187

EvgeniiaVak replied:
Looks the same, but this one would be `spread / bestSell` (not `spread / bestBuy` as in the code), gpt says it can be either.

https://www.wallstreetprep.com/knowledge/bid-ask-spread/
![Screenshot 2023-07-27 at 16 41 19](https://github.com/DeXter-on-Radix/website/assets/27793901/7f442bfd-3568-4d90-88e3-4cd60ecf95e9)

EvgeniiaVak replied:
I like `(2* spread) / (bestBuy+bestSell)` better for the `percentSpread`, @fliebenberg  what do you think?

fliebenberg replied:
Using spread/ask price (best sell) will make your spread % lower, which is probably why many exchanges prefer that. I prefer the more conservative option which is spread/Bid (best buy). The (2 * spread)/bestBuy + bestSell) is a good compromise between the 2. I'd say lets go with that for now. I dont think it is a massively important number (it is more of a general indicator of pair efficiency) and we can always reconsider once we have users using the app and they give us feedback on this.

EvgeniiaVak commented:
![Screenshot 2023-07-27 at 17 28 41](https://github.com/DeXter-on-Radix/website/assets/27793901/acb6fa16-94a9-46d1-93da-ee863a378094)


Probably not an issue with the real data, but on our test data adding K makes it harder to understand the real magnitude of the data change, maybe rounding the decimals, but not replacing 1000s with Ks?

SmashingBumpkin commented:
> ![Screenshot 2023-07-27 at 17 28 41](https://user-images.githubusercontent.com/27793901/256553847-acb6fa16-94a9-46d1-93da-ee863a378094.png)
> 
> Probably not an issue with the real data, but on our test data adding K makes it harder to understand the real magnitude of the data change, maybe rounding the decimals, but not replacing 1000s with Ks?

@EvgeniiaVak I see what you mean. I'm not sure on whats best, perhaps something to go to the community with for feedback?

EvgeniiaVak commented:
@SmashingBumpkin `order-book` is not ready to merge yet, I'll undraft and tell you when it's done. If you want `utils` to be in the `main` branch in the meantime, we can have a separate PR - remove OrderBook.tsx changes from #51 and merge that.

EvgeniiaVak commented:
@fliebenberg in #51 you use `maxDigitsToken1` and `maxDigitsToken2` to determine decimal precision, is that correct? I mean "The maximum number of digits that values of token2 can be specified in." sounds like total number of digits, together with the integers part. This is how it looks like with this setting:
![Screenshot 2023-07-28 at 11 49 34](https://github.com/DeXter-on-Radix/website/assets/27793901/494a9b0b-0943-47f3-ac39-f77daac492e8)

EvgeniiaVak commented:
These `90.15K` and the like make no sense, `90 150` would be much more readable.

SmashingBumpkin reviewed (approved):
Other than the comment on the decimal places this is looking really good. 

I think the significant figure conversation is one we can come back to when we've got a more representative design in place

EvgeniiaVak commented on src/app/OrderBook.tsx:
```diff
@@ -1,30 +1,204 @@
+    props.push({
+      barColor,
+      orderCount: adexRow.noOrders,
+      price: utils.displayNumber(adexRow.price, maxDigitsToken1, true),
```
Thanks! This is not intentional behavior, I even added a test to check - everything works as expected in the displayNumber function. There is a problem - `maxDigitsToken1` from the context gets updated while the code is still in the `toOrderBookRowProps` function, I wonder why... probably initial static state has some values that are different. If so - we need to rerender after the update. On it.
<img width="476" alt="Screenshot 2023-08-01 at 22 33 19" src="https://github.com/DeXter-on-Radix/website/assets/27793901/a136c453-cb33-4ff9-8808-4186e881d1a7">

SmashingBumpkin commented:
On the utils file the logic dictates that numbers above (iirc) 1 get at most 3dp. Replying on my phone so can't check the exact numbersOn Aug 1, 2023 20:37, EvgeniiaVak ***@***.***> wrote:
@EvgeniiaVak commented on this pull request.



In src/app/OrderBook.tsx:
> +  let barColor = "text-green-700";
+
+  if (side === "sell") {
+    iterateStart = adexOrderbookLines.length - 1;
+    iterateEnd = -1;
+    iterateStep = -1;
+    barColor = "text-red-700";
+  }
+
+  for (let i = iterateStart; i !== iterateEnd; i += iterateStep) {
+    const adexRow = adexOrderbookLines[i];
+    total += adexRow.quantityRemaining;
+    props.push({
+      barColor,
+      orderCount: adexRow.noOrders,
+      price: utils.displayNumber(adexRow.price, maxDigitsToken1, true),

Thanks! This is not intentional behavior, I even added a test to check - everything works as expected in the displayNumber function. There is a problem - maxDigitsToken1 from the context gets updated while the code is still in the toOrderBookRowProps function, I wonder why... probably initial static state has some values that are different. If so - we need to rerender after the update. On it.


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

EvgeniiaVak commented:
Ah, yes, I see what you mean, and it doesn't do it for small numbers, fixing.

EvgeniiaVak commented:
@SmashingBumpkin ready for review, I fixed the placement to always be the what's passed in the maxDigitsTokenX variables for small numbers, numbers with K and M now have variable decimals (up to 2), I think it's good that if they are not aligned with the rest of the column - they are now more visible. Also added more tests and fixed the row number to 16.
![Screenshot 2023-08-02 at 13 15 34](https://github.com/DeXter-on-Radix/website/assets/27793901/29c54a72-ce3c-4e63-baf5-9315f45c91e8)

SmashingBumpkin commented:
That's looking consistent now. Good idea on adding tests too.

Another oddity I just noticed is on adex4/xrd. I have a feeling this is coming from adex not the frontend. The max decimals on token1 is 0, so the price is displayed as "0" on all orders, but the orders go down as far as 6dp. Not sure exactly what's going on here, maybe @fliebenberg can assist?

EvgeniiaVak commented:
yes, interesting, it goes from -1 to 0 during the life of the component, I did not expect they could be -1 btw
![Screenshot 2023-08-02 at 21 52 01](https://github.com/DeXter-on-Radix/website/assets/27793901/5b586b94-9630-4e7a-9c34-bc9b1f7e60b3)

EvgeniiaVak commented:
Since we have a little bit more time before the MVP, I found the CSS solution that works in all browsers (by ditching html table and using the grid layout):

EvgeniiaVak commented on src/app/OrderBook.tsx:
```diff
@@ -1,30 +1,219 @@
+    const priceString = utils.displayNumber(price, maxDigitsToken1);
+    const sizeString = utils.displayNumber(size, maxDigitsToken2);
+    const totalString = utils.displayNumber(total, maxDigitsToken1);
+    const barWidth = `${(total / maxTotal) * 100}%`;
```
@fliebenberg @SmashingBumpkin @steferr is that the correct way to calculate the bar width of the orderbook rows?

SmashingBumpkin replied:
Nice one! Looks good. The calculation and display both seem correct.

fliebenberg replied:
Couple of comments:
1. Price is always displayed in terms of token2, so the correct maxDigits for price will be maxDigitsToken2. The way to read the price is: number of token2 I can get / have to pay for one token1. So a price of 100 means I have to pay 100token2 to get 1 token1.
2. Size is usually specified in token1. I.e. it tells you how many token1 is available at that price.
3. The total column is usually just the cumulative sum of the Size columns, so should also be specified in token1. If you want to show the number of token2 in a orderbookline, you can call the column "Value" and then specify the amount of token2 (so it will essentially be the equivalent of the "Size" column, but specified in token2. You could even have a "Total Value" column as well which would just be the "Total" column but specified in token2. This is not typically specified though.
4. Happy with the barWidth calc. Just a question: Is maxTotal the total of displayed lines or all lines (including the ones not displayed)? I think either can work but if I had to choose I would go with displayed lines.

fliebenberg replied:
This is looking great already.
I think we need to start looking at merging some of this work into main, even if it is not completely done. We can then rather do individual issues to improve it. It is just easier when main is not too far out of sync with any particular branch, unless that branch is some new version of the app that we are purposefully keeping separate.

EvgeniiaVak replied:
Ok, fixing the columns, I did have my doubts about which should be which, thank you. `maxTotal` currently only takes into consideration the visible rows. Also there is no scrolling or aggregation, or any other means of getting the rows beyond the first n (currently 8) - which if I recall correctly is what we planned to have in the MVP.
